src/remark/typst.ts

```diff
@@ -0,0 +1,180 @@
+
+const {
+  loaders: { inlineMarkdownImageFileLoader },
+} = getFileLoaderUtils();
+
+const typTemplate = `
+#set page(height: auto, fill: yellow.lighten(95%), margin: 1cm)
+#import "@preview/cetz:0.2.2"\n
+`;
+
+const typcTemplate = [
+  `
+    #set page(height: auto, fill: yellow.lighten(95%), margin: 1cm)
+    #import "@preview/cetz:0.2.2"
+    #align(center, cetz.canvas({
+      import cetz.draw: *
+    `,
+  "\n}))",
+];
+
+const plugin = () => {
+  const transformer = async (ast: MdastRoot, vfile: VFile) => {
+    let children: Promise<void>[] = [];
+    let folder = "typst_renders/";
+    visit(ast, { type: "code" } as const, (node, index, parent) => {
+      if (
+        !(
+          (node.lang === "typ" || node.lang === "typc") &&
+          (node.meta?.includes("render") || node.meta?.includes("example"))
+        )
+      ) {
+        return;
+      }
+
+      if (index === undefined || parent === undefined) {
+        throw new Error("Invalid location for typst code example");
+      }
+
+      const hash = createHash("md5")
+        .update(node.value)
+        .digest("hex")
+        .slice(0, 6);
+      const path = folder + hash + ".svg";
+      if (!existsSync(path)) {
+        children.push(
+          new Promise((resolve) => {
+            const child = exec("typst c - " + path);
+
+            if (!child.stdout || !child.stderr || !child.stdin)
+              throw new Error(`Failed to spawn typst process`);
+
+            child.stdout.pipe(process.stdout);
+            child.stderr.pipe(process.stderr);
+            if (node.lang === "typ") {
+              child.stdin.write(typTemplate + node.value);
+            } else {
+              child.stdin.write(typcTemplate[0] + node.value + typcTemplate[1]);
+            }
+            child.stdin.end();
+            child.on("exit", () => resolve());
+          })
+        );
+      }
+
+      if (node.meta.includes("render")) {
+        parent.children[index] = {
+          type: "image",
+          url: "@site/" + path,
+        };
+      } else {
+        // from https://github.com/facebook/docusaurus/blob/aab332c2ae3f78df5cdac6344bab0075ea07a035/packages/docusaurus-mdx-loader/src/remark/transformImage/index.ts#L45-L118
+        let relativeImagePath = posixPath(relative(dirname(vfile.path), path));
+        relativeImagePath = `./${relativeImagePath}`;
+        const requireString = `${inlineMarkdownImageFileLoader}${escapePath(
+          relativeImagePath
+        )}`;
+        parent.children[index] = {
+          type: "mdxJsxFlowElement",
+          name: "TypstPreview",
+          attributes: [
+            {
+              type: "mdxJsxAttribute",
+              name: "imgUrl",
+              value: assetRequireAttributeValue(requireString),
+            },
+            {
+              type: "mdxJsxAttribute",
+              name: "isBig",
+              value: boolToJsx(node.meta?.includes("big")),
```

Comment:
Can we use the term "stack" or "vertical" instead of "big". We use the same term in the pdf manual for the examples

Reply:
It's "vertical" now to be the same as in the pdf manual.